### PR TITLE
Apertures

### PR DESCRIPTION
**Pull request by Jordan-Dennis**
Hi all,
the long awaited and delayed pull request is finally here. I'm opening it as a draft as there are large amounts that will not be changing and can be reviewed now. Things that will be changing are documented in issues #140, #141, #142 and #143. Moreover, the final touches need to be made to the tests and I will make a detailed tutorial.

### CODE REVIEW

**Comment by LouisDesdoigts** [comment]
Yah this is big one 😅 - I can start to review this but will likely take a while, will be focusing on getting paper 1 submitted today. 

What are the main files to review?

**Comment by Jordan-Dennis** [comment]
`apertures.py` contains it all. There is only about `3000` lines that survived.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -31,25 +30,31 @@ jobs:
...
       - name: optics tests
-        run: pytest --quiet tests/optics.py
+        run: pytest --quiet tests/test_optics.py
 
       - name: detector tests
-        run: pytest --quiet tests/detectors.py
+        run: pytest --quiet tests/test_detectors.py
+
+      - name: aperture tests
+        run: pytest --quiet tests/apertures.py
```
Comment (LouisDesdoigts) [inline .github/workflows/tests.yml:54]: should be test_apertures.py?

File: dLux/apertures.py [hunk 2]
```diff
@@ -0,0 +1,2641 @@
+import equinox as eqx
+import matplotlib as mpl 
+import matplotlib.pyplot as plt
+import jax.numpy as np
+import jax 
+import dLux
+import abc
+import functools
+from typing import TypeVar 
+
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:10]: So we want to import specific functions or classes from packages, not the whole package (check the CONTRIBUTING.md file)

File: dLux/apertures.py [hunk 3]
```diff
@@ -0,0 +1,2641 @@
...
+__all__ = ["factorial", "CircularAperture", "SquareAperture", 
+    "HexagonalAperture", "RegularPolygonalAperture", 
+    "IrregularPolygonalAperture", "StaticAperture",
+    "AberratedAperture", "StaticAberratedAperture"]
+
+
+two_pi: float = 2. * np.pi
+        
+ 
+def factorial(n : int) -> int:
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:25]: This really should be in the utils folder as a standalone function

Comment (LouisDesdoigts) [reply dLux/apertures.py:25]: Also I though the whole point was that this works for floats, not just integers? type hints and docs need updating if so

File: dLux/apertures.py [hunk 4]
```diff
@@ -0,0 +1,2641 @@
...
+    The ApertureLayer groups together all of the functionality 
+    that is associated with the apertures. Very little of this
+    functionality is actually implemented by itself because 
+    implementation varies amongst the subclasses. It is a 
+    layer within the class Heirachy that exists almost purely 
+    as a classification.
+
+    Parameters:
+    -----------
+    name: String
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:54]: Use the python type, ie `str` not `String`

File: dLux/apertures.py [hunk 5]
```diff
@@ -0,0 +1,2641 @@
...
+        system. 
+    """
+
+    
+    def __init__(self: dLux.optics.OpticalLayer) -> dLux.optics.OpticalLayer:
+        """
+        Automatically assigns the name of the layer to be the 
+        class name. 
+        """
+        self.name = self.__class__.__name__
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:65]: This is a nice little trick, I have a q though: Will this instantiate the name of the lowest or highest level class? Ie will this become `HexagonalAperture` or `ApertureLayer` one this is inherited?

File: dLux/apertures.py [hunk 6]
```diff
@@ -0,0 +1,2641 @@
...
+class AbstractDynamicAperture(ApertureLayer, abc.ABC):
+    """
+    AbstractDynamicAperture:
+    ------------------------
+    This class also provides a level of classification without 
+    implementing any functionality. It was created so that 
+    `DynamicAberratedAperture` and `DynamicAperture` could 
+    inherit from a common base. 
+
+    Parameters:
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:77]: Attributes not parameters

File: dLux/apertures.py [hunk 7]
```diff
@@ -0,0 +1,2641 @@
...
+class ApertureLayer(dLux.optics.OpticalLayer, abc.ABC):
+    """
+    The ApertureLayer groups together all of the functionality 
+    that is associated with the apertures. Very little of this
+    functionality is actually implemented by itself because 
+    implementation varies amongst the subclasses. It is a 
+    layer within the class Heirachy that exists almost purely 
+    as a classification.
+
+    Parameters:
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:52]: attributes not parameters

File: dLux/apertures.py [hunk 8]
```diff
@@ -0,0 +1,2641 @@
...
+    AbstractDynamicAperture:
+    ------------------------
+    This class also provides a level of classification without 
+    implementing any functionality. It was created so that 
+    `DynamicAberratedAperture` and `DynamicAperture` could 
+    inherit from a common base. 
+
+    Parameters:
+    -----------
+    centre: float, meters
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:79]: Array, meters,
the (x,y) coordinates of the center of the aperture.

File: dLux/apertures.py [hunk 9]
```diff
@@ -0,0 +1,2641 @@
...
+        The x coordinate of the centre of the aperture.
+    strain: Array
+        Linear stretching of the x and y axis representing a 
+        strain of the coordinate system.
+    compression: Array 
+        The x and y compression of the coordinate system. This 
+        is a constant. 
+    rotation: float, radians
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:87]: Array, not float
Which direction? clockwise or counter clockwise?

File: dLux/apertures.py [hunk 10]
```diff
@@ -0,0 +1,2641 @@
...
+    compression: Array
+    rotation: Array
+    
+
+    def __init__(
+            self        : ApertureLayer, 
+            centre      : Array = [0., 0.], 
+            strain      : Array = [0., 0.],
+            compression : Array = [1., 1.],
+            rotation    : Array = 0.) -> ApertureLayer:
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:102]: I would wrap all these default inputs in np.arrays, in order to have them match their type hints

File: dLux/apertures.py [hunk 11]
```diff
@@ -0,0 +1,2641 @@
...
+        """
+        The default aperture is dis-allows the learning of all 
+        parameters. 
+
+        Parameters
+        ----------
+        centre: float, meters
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:109]: Array, (x,y)

File: dLux/apertures.py [hunk 12]
```diff
@@ -0,0 +1,2641 @@
...
+            The centre of the coordinate system along the x-axis.
+        softening: bool = False
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:111]: No softening input

File: dLux/apertures.py [hunk 13]
```diff
@@ -0,0 +1,2641 @@
...
+            True if the aperture is soft edged otherwise False. A
+            soft edged aperture has a small layer of non-binary 
+            pixels. This is to prevent undefined gradients. 
+        occulting: bool = False
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:115]: No occulting input

File: dLux/apertures.py [hunk 14]
```diff
@@ -0,0 +1,2641 @@
...
+        self.centre = np.asarray(centre).astype(float)
+        self.strain = np.asarray(strain).astype(float)
+        self.compression = np.asarray(compression).astype(float)
+        self.rotation = np.asarray(rotation).astype(float)
+
+
+    def _rotate(self: ApertureLayer, coords: Array) -> Array:
+        """
+        Rotate the coordinate system by a pre-specified amount,
+        `self._theta`
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:138]: `self.rotation`

Comment (LouisDesdoigts) [inline dLux/apertures.py:138]: self.rotation

**Comment by LouisDesdoigts** [comment]
I think it might make sense to move the coordinate transforming methods into the utils/coordinates.py script

File: dLux/apertures.py [hunk 15]
```diff
@@ -0,0 +1,2641 @@
...
+            softening   : bool = False) -> ApertureLayer:
+        """
+        The default aperture is dis-allows the learning of all 
+        parameters. 
+
+        Parameters
+        ----------
+        centre: float, meters
+            The centre of the coordinate system along the x-axis.
+        softening: bool = False
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:300]: This is input as a bool but saved as a float

File: dLux/apertures.py [hunk 16]
```diff
@@ -0,0 +1,2641 @@
...
+    strain: Array
+        Linear stretching of the x and y axis representing a 
+        strain of the coordinate system.
+    compression: Array 
+        The x and y compression of the coordinate system. This 
+        is a constant. 
+    rmax : float
+        The proportion of the pixel vector that is contained within
+        the outer ring of the aperture.
+    rmin : float
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:486]: rmin, rmax are `Array` type

File: dLux/apertures.py [hunk 17]
```diff
@@ -0,0 +1,2641 @@
...
+    softening: bool 
+        True if the aperture is soft edged otherwise False. A
+        soft edged aperture has a small layer of non-binary 
+        pixels. This is to prevent undefined gradients. 
+    occulting: bool 
+        True if the aperture is occulting else False. An 
+        occulting aperture is zero inside and one outside. 
+    """
+    rmin : float
+    rmax : float
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:498]: Arrays

File: dLux/apertures.py [hunk 18]
```diff
@@ -0,0 +1,2641 @@
...
+            rmin        : Array, 
+            centre      : Array = [0., 0.],
+            strain      : Array = [0., 0.],
+            compression : Array = [1., 1.],
+            occulting   : bool = False, 
+            softening   : bool = False) -> ApertureLayer:
+        """
+        Parameters
+        ----------
+        centre: float, meters
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:512]: Array

**Comment by LouisDesdoigts** [comment]
In the class docsting we describe the Attributes, not the parameters, parameters describe function/method inputs

File: dLux/apertures.py [hunk 19]
```diff
@@ -0,0 +1,2641 @@
...
+        non_occ_ap: Array 
+            This is essential the final step in processing to produce
+            the aperture. What is returned is the non-occulting 
+            version of the aperture. 
+        """
+        bc_x1: float = self.vertices[:, 0][:, None, None]
+        bc_y1: float = self.vertices[:, 1][:, None, None]
+
+        bc_x: float = coords[0][None, :, :]
+        bc_y: float = coords[1][None, :, :]
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:1311]: I would just add some one-liners describing what `bc_x1` are etc. It's a little hard to see _what_ this function is doing with these somewhat abstract names

File: dLux/apertures.py [hunk 20]
```diff
@@ -0,0 +1,2641 @@
...
+            
+        wedges: float = self._make_wedges(phi, bounds)
+        ms: float = -1 / np.tan(2. * i * alpha + alpha)
+        xs: float = self.rmax * np.cos(2. * i * alpha)
+        ys: float = self.rmax * np.sin(2. * i * alpha)
+        dists: float = self._perp_dists_from_lines(ms, xs, ys, x, y)
+        inside: float = self._is_orig_left_of_edge(ms, xs, ys)
+         
+        dist: float = (inside * dists * wedges)
+        return self._soften(dist.sum(axis=0))
```
Comment (LouisDesdoigts) [inline dLux/apertures.py:1470]: This function uses the different syntax of val : dtype = ... - I would remove these just to keep consistency across the package

**Comment by LouisDesdoigts** [comment]
So I've done a bit of a quick review of part of the apertures.py file. It all looks really good, for the most part it just needs a thorough once-over of the docsrings and type hints. Happy for this to be left till the end, so that things can keep changing if need be in the dev stage. 

The code all looks really good though, only real thing is to add some comments as to _what_ is being done in some of the bigger methods as they have some abstract variable names and so can be a little hard to follow what is happening. 

I'm happy if you just want to focus on the tests and code, and I can do a thorough once over of the docstrings etc after that!

File: .github/workflows/tests.yml (hunk 1 above)
Comment (Jordan-Dennis) [reply .github/workflows/tests.yml]: :+1:

File: dLux/apertures.py (hunk 2 above)
Comment (Jordan-Dennis) [reply dLux/apertures.py:10]: How does this work with `import jax.numpy as np`. Are we switching to `from jax import numpy as np`?

File: dLux/apertures.py (hunk 3 above)
Comment (Jordan-Dennis) [reply dLux/apertures.py:25]: Factorial is only defined mathematically for integers. The gamma function is defined for floats. The use of `int` was purposeful, but you are correct, the implementation is agnostic.

File: dLux/apertures.py (hunk 8 above)
Comment (Jordan-Dennis) [reply dLux/apertures.py:79]: Will through a reference to #138. I have to go through multiple places and change this. With inherited documentation that would not be the case.

File: dLux/apertures.py (hunk 15 above)
Comment (Jordan-Dennis) [reply dLux/apertures.py:280]: Yes. That is on purpose, see #142.

**Comment by Jordan-Dennis** [comment]
OK @LouisDesdoigts things should be good for you to go through. There will be some minor changes tomorrow from #142 but other than that it is just the documentation getting added to the `docs` folder along with a tutorial.

**Comment by LouisDesdoigts** [comment]
A few comments/questions, though this would be easier than sporadic slack messages and spare comments in the PR:

 - The high level `Aperture` class should call `super().__init__(name=name)` and take name in as a kwarg
 - Optimise the `._aperture` and `._basis` methods with `jax.tree_map` -> See `core.py: model()` for examples
 - set `Compound/Composite/MultiAperture` to take in a list and use the `list_to_dict` method to initialise the internal dictionary
 - Create a `.get_aperture(diameter, npixels)`, `.get_basis(diameter, npixels)` method in `Aperture` that calculates the aperture and basis arrays for the purposes of plotting/visualisation
 - Enable nan checks (via the jax debug nans flag) in the tests
 - Move the basis generation functions into utils. Its a really awesome set of functions that I think should be stand alone. This is low priority though, would be happy to merge the pr first and set that up later, if yall prefer

I think once we are happy with all this, I'll do a big once over for docs/formatting (theres still some stray `Matrix`, `Tensor` etc and some of the `var: dtype = ...`)
 
 Reminder for myself: Implement a `__getattr__` method and clue in higher level classes to call this method

**Comment by LouisDesdoigts** [comment]
Oh and as a fast way to do jit and higher order grad tests, try adding some complex aperture layers into the tutorial notebooks and make sure things don't break

**Comment by Jordan-Dennis** [comment]
What does the second point mean? Is that for the `MultiAperture` and `CompoundAperture`? There is nothing to map over for the individual aperture classes (that I am aware of). As for 1 and 3, this is why we need a change log which I will implement as part of the pull request.

**Comment by Jordan-Dennis** [comment]
I plan to keep @ataras2's `has_aberrated` flag as it let's us make the small optimisation of only calculating the path differences if an `AberratedAperture` was included very easily.

**Comment by Jordan-Dennis** [comment]
OK, that should do it. The PR is ready for it's final review.

**Comment by LouisDesdoigts** [comment]
Alright so just a dip into the apertures branch, made a bunch of docstring fixes and updates, a few namespace changes, just a lot of minor but important changes. It's all looking very good and I have a much better idea of everything at this point. So there are a few last things that need to happen for before the merge in (vague) importance order.

- [ ] Fix the apertures tests and re-enable the jax debug nans flag. This is the most important, we never want any nans in our code. The last test of the apertures produces a bunch of nans and infs which shouldn't happen. It looks like it originates from the basis stuff, but I'm not sure beyond that. 
- [ ] Adding assert statements to constructors. There are a lot of constructors that don't test things like dimensionality. This is is important as you should not be able to construct an object in dLux that then throws an error when it runs, and it should give helpful responses as to the correct formatting for some input (ie, rotation should be a zero-dim array). This shouldn't be too hard, just annoying. The tests should be updated to test these. 
- [ ] Fix phase/opd inconsistency in the aberrations. Some add phase, some add opd, this should actually be an *option* for the user. Easiest is probably to default to applying an opd with a simple boolean class attribute that is checked in the `__call__` method.
- [ ] Reduce the number of pixels used to test the apertures class. The tests take too long, I would say using anything larger than a 16x16 array is overkill.
- [ ] Add some softening amount parameter. As far as I can tell this should just be a simple value that is multipled in the `._soften()` methods.
- [ ] Fix/Update the tutorial notebooks to work with the new code. Some of the notebooks rely on deprecated layers like the old `CircularAperture`. These should be updated to use the new layers (I am happy to fix these up though).
- [ ] Clean up the branch. Namely in the dev folder, specifically the `jwst.py`, `orthonormalisation.py`, and the `tests/` directory. If there is important code it should be moved into the relevant scripts, else it should be moved into a different branch and then be removed for the PR.

After these changes then we should be able to merge into main. There is one last sticking point though about these basis stuff. I still maintain that the basis generation (and aperture generation stuff tbh) should be moved into scripts in the utils folder. The core is just jax functions that are very useful outside of the apertures and it doesn't make sense to me that you should need to create an aperture class to pre-calculate some polike basis for use elsewhere. I am aware this ties into the wider discussion (#157) so we can discuss that further there, and I would be more than happy to merge the PR with things where they are now.

Anyway this is all so good, I had a bit of a play around with the classes and I think the API is 🤌🤌. Very much looking forward to using this in the future, its awesome!

**Comment by Jordan-Dennis** [comment]
You do realise that you were getting annoyed when I had a floating point `softening` parameter :laughing: :rofl:.

**Comment by Jordan-Dennis** [comment]
"Fix phase/opd inconsistency in the aberrations."

It only works through OPD at the moment. Where are you referring to inconsistency?

**Comment by LouisDesdoigts** [comment]
> Where are you referring to inconsistency? 

The call method of the `StaticAberratedAperture` uses `.add_phase`. 

I also just noticed a few things about that class:
1. It should inherit from the `StaticAperture` class (not sure how this fits with the wider inheritance structure, but it duplicates the `.aperture` parameter), it would make sense to inherit from `StaticAperture`.
2. The call method applied the `.method().method()` to the wavefront - can be improved with an `eqx.tree_at` to apply the updates simulatensously.
3. It just applies the basis, rather than calculating the opd from the `coefficients` multiplied to the `basis`.

**Comment by LouisDesdoigts** [comment]
> You do realise that you were getting annoyed when I had a floating point softening parameter 😆 🤣.

Yeah I had a misunderstanding of how the method worked at the time, I thought it applied a single pixel threshold, rather than a multi-pixel threshold. Since its not actually a binary single pixel threshold, but actually a continuous parameter. It does make sense to have a float since there are cases where we might be working with very large or small optics in which case it would be highly valuable to set this manually.

I've been meaning to ask, from what I can glean from the code it looks like the threshold amount is in *physical* units (ie meters). It would make sense to me to have the parameter in units of pixels, not sure if that would be very difficult to change though.

**Comment by Jordan-Dennis** [comment]
So to get it in units of pixels we can just divide by the pixel scale. The steepness parameter actually has units of, inverse distance. This is so that `np.tanh` has a unit-less parameter. Setting the number of soft pixels is a fine art that **there is no exact answer for**. The way I currently have it set up produces about 2-3 pixels that are *obviously* non-binary. Because the function is continuous there are many more that are not *very slightly* not binary. We can always exchange the `np.tanh` for a custom piece-wise continuous function. I think that if you want to explicitly chose the number of pixels, this would be the way to go as it gives an exactly linear response. I'm not sure how this would interact with the gradients though since, just a hard edged (binary) aperture is technically also piece-wise continuous.

**Comment by Jordan-Dennis** [comment]
I have encountered a familiar problem with have a **public** (formerly **private**) parameter to control the width of the soft edged region. Namely, *the magnitude of this parameter depends on the pixel scale*. This means that setting a default value at `__init__` time is not very easy. I have found that a good (heuristically speaking) value is the number of pixels along on edge of the image (2-3 *soft pixels*). Since this is not known at `__init__` time except in the static case (coordinates are in the `Wavefront` and the `StaticAperture`s). 

The solution I have come up with involves a `lax.cond`/`if` along the lines of `if not self.steepness: then ...`. This is not ideal because we run into all the usual `jit` problems. Does `equinox` automatically make such arguments behave like `static_args` or can we use `eqx.static_field` to achieve this?

**Comment by Jordan-Dennis** [comment]
This has been resolved via a heuristic command and relaxing the exactness.

**Comment by Jordan-Dennis** [comment]
OK, so as we suspected the `nan`s were occurring in the extra memory that get's allocated to make the basis code compile. They were the result of dividing infinite values by infinite values. The infinite values were in turn resulting from the gamma function that powers the factorial getting evaluated on negative numbers. I wonder if we should add a flag to `factorial`?

With this comment the code is ready for its final review @LouisDesdoigts. All you should need to do is check the tutorial notebooks have the correct output since I cannot run them on my computer.

**Comment by Jordan-Dennis** [comment]
So I just tried to boot my forwards model again and there is a `ConcretisationError` that occurs now with a zero dimensional array and `!= 0.`. I am stopping for the week but I will make this the first point of call on Tuesday.

**Comment by Jordan-Dennis** [comment]
The last 115 commits changed nothing in the actual code (really) and were mostly my examples and attempts at optimisation.

**Review by LouisDesdoigts — APPROVED** [review]
Ready for merge with main
